Untangle the section id, factories and modules

atenart opened the pull request:
Semi-RFC material; since code is ready let's not flag it but feel free to discuss the approach :)

The main goal of this series is $title, and some preliminary work for a possible future removal of the module concept. In any way this solve issues linked to how the section ids were used which did not map to the reality:

- Factories are tied to a BPF counterpart, not to an event section. Eg. the tracking section is generated at post-processing time and does not have any link to the BPF side.
- Modules are not strictly tied to event sections. Eg. all core sections do not have a corresponding module.
- Factories and modules are not in a 1:1 relationship. Eg. core factories are not in a module.

This also now allows to have a factory producing more than a single type of event section; this is no side effect and is intended.

Overall this was something I wanted to tackle for some time now, but I was keeping it for later. But adding support for `enum sk_rst_reason` showed w/o this preliminary rework things would have gotten even worse between event sections, factories and modules. This is why I chose to keep those patches in that PR; but I can split into two PRs if needed.

One question I have is with the last commit: should be put all those "common types" into the same event section or add one for each. Having a dedicated section helps for possible future improvement (more fields when applicable), but makes event section count to potentially grow quite a lot. Also I should find a better name instead of "common types".

atenart commented:
Again we have a failure due to benchmarking being so difficult to maintain. I'm not going to fix this here as we have #412; so let's merge it first.

amorenoz commented:
Mainly I have comments on the last two patches. For the code refactorization, besides nits, it looks great to me and I think it goes in the right direction of getting rid of the module abstraction.

atenart commented on retis-events/src/ct.rs:
```diff
@@ -112,7 +112,8 @@ pub struct CtEvent {
 /// Conntrack event
-#[event_section("ct")]
+#[event_type]
+#[derive(Default)]
```
Right :)

atenart commented on retis-derive/src/lib.rs:
```diff
@@ -10,18 +10,21 @@ pub fn event_section(
 
         impl #ident {
-            pub(crate) const SECTION_NAME: &'static str = #name;
+            pub(crate) const SECTION_ID: u8 = #id as u8;
```
I think that's fine. We already do this, https://github.com/retis-org/retis/blob/main/retis/src/core/events/bpf.rs#L669

If we want something a bit more explicit, we could reuse the raw value but use a dedicated "variant". Haven't tested the following code, but might look like it:

```rust
pub enum SectionId {
    Common,
    ...
}

impl SectionId {
    pub const Test: Self = Self::Common;
}
```

amorenoz commented on retis/src/core/probe/kernel/bpf/include/common.h:
```diff
@@ -282,6 +283,8 @@ static __always_inline int chain(struct retis_context *ctx)
 	barrier_var(pass_threshold);
 
+	handle_common_types(ctx, event);
+
```
> > I think this might be problematic. Not all functions (and therefore not all probes) that have `struct skb_event_reason` as argument are actually dropping the packet.
> 
> Right.
> 
> > We could very easily end up having duplicate drop events that would be terribly confusing. I do think `DropEvent` must come _only_ from `kfree_skb`. Am I missing a use-case?
> > I can see some kind of generic "tracking type" or "type-driven events" for some interesting types but we should be very clear that it just means "this probe happened to have this type and value in its arguments". IMO, A DropEvent has a very different meaning.
> 
> Drops are quite complex. Bottom line is drop reasons are not linked to a packet effectively being dropped: eg. the kernel could be too old (in case we emulate a reason, but I think we should stop doing so) or the reason might be `SKB_COMSUMED` or `SKB_NOT_DROPPED_YET`. Detecting a packet is being dropped is really about seeing it in the right tracepoint: `skb:kfree_skb`; that's all[1]. The `skb-drop` section was just here to hold drop-related information, not that the packet was effectively dropped.
> 
> Now, I agree this is confusing and not a great UX. The naming of `skb-drop` was pretty bad.
> 
> Then on why retrieving those two "common types". More and more functions in the kernel accept those "reason" enums as parameters and are even returning them. Retrieving those values gives very valuable information, as we can understand better the flow of a packet, and how a function was behaving.
> 

Can you give a concrete example?


> Finally (I think that's important for finding a solution) I wasn't sure about how to model those "common types" in the events. I chose a single section per type (and reused `skb-drop` for this), but that might not be the perfect solution.
> 
> I think we should improve both situations.
> 
> One possibility: remove the `skb-drop` section and provide a `drop()` helper for events. A bit like the `tracking_id()` helper in the tracking section. That new helper could be used to flag a packet being dropped in the output, and later in Python. It would just be hidden from the raw JSON event (IMO that is fine). In addition the drop reason should be stored in a new section. The `skb-drop` collector can stay, it has value in probing at the right place.
> 

Maybe that's an option but my feeling is that "packets being dropped" is  way more important than "functions happen to pass or return some drop reason". I think the UX should strongly favor the former and try by all means not to get them mixed.

How about:
- We keep the `DropEvent` linked uniquely to kfree_skb. We make it clear in the string representation that there was a packet drop.
- We add a `SkResetEvent` linked uniquely to trace. We make it clear in the string representation that a rst was actually sent.
- We have a `ProbeArgsEvent` or `ProbeMetadata` or `ProbeInfo` (something like that) that gives information about the probe itself, its arguments and return code for known types. We could support native C types and some well-known structs or enums. Here we can add skb_drop_reason but the string representation should be super-clear this is an argument of the function, e.g: 

```
 [k] [tcp_rcv_state_process] skb ffff8e8ec1ab3300
   probe info: arg0(struct sock*) = 0x123445, arg1(struct skb*) = 0xffff8e8ec1ab3300, ret(enum skb_drop_reason) = SKB_DROP_REASON_TCP_INVALID_SEQUENCE
   192.168.1.137 > 192.168.1.138 ttl 64 tos 0x0 id 49968 off ... rest of skb section ...
   [... other sections ...]

 [k] [kfree_skb_reason] skb ffff8e8ec1ab3300
   192.168.1.137 > 192.168.1.138 ttl 64 tos 0x0 id 49968 off ... rest of skb section
    drop (reason TCP_OLD_SEQUENCE) 
```


> Two additional comments:
> 
>     * I forgot to remove the `skb-drop` section generation from the `skb-drop` collector. It is now handled by the core.
> 
>     * We can split the generic types part from the rest of the series, so we can merge the initial rework.
> 

I definitely think we should as the rest of the PR is almost ready to get merged and can unblock future work (in fact python series would need an interesting  rebase).

> 
> [1] It's in fact even more complex, there are different definitions of what is a "drop". Eg. run `retis -p generic collect -f icmp --cmd 'ping 1.1.1.1'` and watch packets being "drops" :)

I see no drops here, but in any case, even if they are, this is the kernel reporting drops, we cannot do better than the kernel. But we should do worse (i.e: making non-drops feel like drops).

atenart commented:
Rebased on top of #411 (as it seems to be ready) to avoid conflicts and removed the common type commits from this. I'll open a dedicated PR and we can have the discussion about those there.

atenart commented:
Same CI error as the others, this week our running CI can't work.

atenart commented:
Removed the same commit as in #411.

atenart commented:
The benchmarking CI is actually failing for a good reason; will have a look.

atenart commented:
All good now.

amorenoz commented on retis-derive/src/lib.rs:
```diff
@@ -79,13 +78,29 @@ pub fn event_type(
+        }
+
         impl EventSectionFactory for #ident {
+            fn factory_id(&self) -> u8 {
```
nit: I think we can remove the "factory" part as it's redundant. The method must be called on a  factory.

atenart replied:
Same with `section_id()` for event sections.

amorenoz reviewed (dismissed):
Leaving a nit aside, the PR LGTM.

atenart commented:
Renamed `section/factory_id()` helpers into `id()`.

atenart commented:
Rebased on top of main to fix a conflict. No other change.
